core/src/main/java/org/infinispan/container/versioning/irac/DefaultIracTombstoneManager.java

```diff
@@ -277,36 +279,58 @@ private CompletionStage<Void> performCleanup() {
       if (stopped) {
          return CompletableFutures.completedNull();
       }
-      boolean trace = log.isTraceEnabled();
+      var trace = log.isTraceEnabled();
 
       if (trace) {
          log.trace("[IRAC] Starting tombstone cleanup round.");
       }
 
       scheduler.onTaskStarted(tombstoneMap.size());
 
-      CompletionStage<Void> stage = Flowable.fromIterable(tombstoneMap.values())
+      var tombstoneStage = Flowable.fromIterable(tombstoneMap.values())
             .groupBy(this::classifyTombstone)
             // We are using flatMap to allow for actions to be done in parallel, but the requests in each action
             // must be performed sequentially
-            .flatMap(group -> {
-               switch (group.getKey()) {
-                  case REMOVE_TOMBSTONE:
-                     return removeAllTombstones(group);
-                  case NOTIFY_PRIMARY_OWNER:
-                     return notifyPrimaryOwner(group);
-                  case CHECK_REMOTE_SITE:
-                     return checkRemoteSite(group);
-                  case KEEP_TOMBSTONE:
-                  default:
-                     return Flowable.empty();
-               }
+            .flatMap(group -> switch (group.getKey()) {
+               case REMOVE_TOMBSTONE -> removeAllTombstones(group);
+               case NOTIFY_PRIMARY_OWNER -> notifyPrimaryOwner(group);
+               case CHECK_REMOTE_SITE -> checkRemoteSite(group);
+               default -> Flowable.empty();
             }, true, ACTION_COUNT, ACTION_COUNT)
             .lastStage(null);
       if (trace) {
-         stage = stage.whenComplete(TRACE_ROUND_COMPLETED);
+         tombstoneStage = tombstoneStage.whenComplete(TRACE_ROUND_COMPLETED);
       }
-      return stage.whenComplete(scheduler);
+
+      var keysStage = Flowable.fromStream(iracManager.running().pendingKeys())
+            .groupBy(IracManagerKeyInfo::segment)
+            .filter(group -> isNotPrimaryOfSegment(group.getKey()))
+            .flatMap(group -> group
+                        .buffer(batchSize)
+                        .concatMapDelayError(this::checkKeysExistsInPrimary),
+                  true, segmentCount, segmentCount)
```

Comment:
You must subscribe to every GroupedFlowable otherwise it can hang the operation (the filter method isn't subscribing). 

You need to do something like

```java
    .flatMap(group -> {
       if (isNotPrimaryOfSegment(group.getKey())) 
         return group
                        .buffer(batchSize)
                        .concatMapDelayError(this::checkKeysExistsInPrimary);
       else 
         return group.ignoreElements();
    }, true, segmentCount, segmentCount)
```

But you were correct in that you needed to provide the concurrency level to be the maximum number of segments, otherwise it could have hung as well.